stmhal/sdcard.c

```diff
@@ -85,6 +86,29 @@ void sdcard_init(void) {
     // invalidate the sd_handle
     sd_handle.Instance = NULL;
 
+    // configure the SD card detect pin
+    // we do this here so we can detect if the SD card is inserted before powering it on
+    GPIO_Init_Structure.Mode = GPIO_MODE_INPUT;
+    GPIO_Init_Structure.Speed = GPIO_SPEED_LOW;
+#if defined(MICROPY_HW_SDCARD_DETECT_PIN)
+    // assume a mechanical switch is used
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = MICROPY_HW_SDCARD_DETECT_PIN.pin_mask;
+    HAL_GPIO_Init(MICROPY_HW_SDCARD_DETECT_PIN.gpio, &GPIO_Init_Structure);
+#elif defined(MICROPY_HW_SDCARD_DETECT_PULL)
+    // assume no mechanical switch, so assume CD/DAT3 is used for card detect.
+    GPIO_Init_Structure.Pull = MICROPY_HW_SDCARD_DETECT_PULL;
+    GPIO_Init_Structure.Pin = pin_C11.pin_mask;
+    HAL_GPIO_Init(pin_C11.gpio, &GPIO_Init_Structure);
+    // capture sdcard present state before the IO pin is remapped to SDIO functionality
+    sdcard_present = (HAL_GPIO_ReadPin(pin_C11.gpio, pin_C11.pin_mask) == MICROPY_HW_SDCARD_DETECT_PRESENT);
+    // TODO: do we need to issue ACMD42 (SD_CMD_SD_APP_SET_CLR_CARD_DETECT) to disable the internal pullup resistor?
+#else
+    // SD Card Detect not defined.  Just assume it's there.
```

Comment:
SD cards have a 50k (nominal) pull up resistor on CD/DAT3.  The Olimex E407 has a 1Mohm (weak) pull down on the CD/DAT3.  So it reads low if no SD card inserted, and high if inserted.

I'm guessing the teensy I/O is configurable (yes/no?) and could be configured to be an input with inbuilt pulldown (yes/no?).  Not sure what the value of the inbuild pullup/down resistors are.

The K66 reference manual states the following re the DAT3.

> The SD standard requires the SD bus signals (except the SD clock) to be pulled up during
> data transfers. The SDHC also provides a feature of detecting card insertion/removal, by
> detecting voltage level changes on DAT[3] of the SD bus. To support this DAT[3] must
> be pulled down. To avoid a situation where the SDHC detects voltage changes due to
> normal data transfers on the SD bus as card insertion/removal, the interrupt relating to
> this event must be disabled after the card has been inserted and detected. It can be reenabled after the card is removed.